commit a56110a3e30942e8d374c45e4cd6a7d67cc274db (refs/changes/31/931/3)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-11-22T00:18:47+00:00 (2 years, 11 months ago)
    
    joyent/node-cueball#52 possible for claim() to give out a connection that has suddenly died
    Reviewed by: Cody Peter Mello <cody.mello@joyent.com>

diff --git a/lib/connection-fsm.js b/lib/connection-fsm.js
index 5f778e9..e15d01f 100644
--- a/lib/connection-fsm.js
+++ b/lib/connection-fsm.js
@@ -157,14 +157,17 @@ ConnectionFSM.prototype.claim = function (stack, cb) {
 	    map(function (l) { return (l.replace(/^[ ]*at /, '')); });
 	this.cf_claimed = true;
 	var self = this;
-	this.once('stateChanged', function (st) {
-		if (st === 'busy') {
+	this.on('stateChanged', onStateChanged);
+	function onStateChanged(st) {
+		if (st === 'busy' && self.isInState('busy')) {
+			self.removeListener('stateChanged', onStateChanged);
 			cb(null, self.cf_shadow, self.cf_conn);
-		} else {
+		} else if (st !== 'busy') {
+			self.removeListener('stateChanged', onStateChanged);
 			cb(new Error('Claimed connection entered state ' +
 			    st + ' during claim, instead of "busy"'));
 		}
-	});
+	}
 	this.emit('claimAsserted');
 };
 
